contracts/ColonyFunding.sol

```diff
@@ -178,13 +175,23 @@ contract ColonyFunding is ColonyStorage, DSMath {
   }
 
   function claimRewardPayout(uint256 _payoutId, uint256[7] _squareRoots, uint256 _userReputation, uint256 _totalReputation) public {
+    require(!userRewardPayouts[msg.sender][_payoutId], "colony-reward-payout-already-claimed-or-waived");
+
     RewardPayoutCycle memory payout = rewardPayoutCycles[_payoutId];
+    // Checking if payout is active
     require(block.timestamp - payout.blockTimestamp <= 60 days, "colony-reward-payout-not-active");
-    require(_payoutId - userRewardPayoutCount[msg.sender] == 1, "colony-reward-payout-bad-id");
 
-    //TODO: Prove that userReputation and totalReputation in reputationState are correct
+    //TODO: Prove that userReputation and totalReputation in reputationState (reputation root has at the start of the payout) are correct
 
-    uint256 userTokens = token.balanceOf(msg.sender);
+    ITokenLocking tokenLocking = ITokenLocking(IColonyNetwork(colonyNetworkAddress).getTokenLocking());
+
+    uint256 userLockCount;
+    uint256 lastDepositTimestamp;
+    (userLockCount, lastDepositTimestamp) = tokenLocking.getUserTokenLockInfo(address(token), msg.sender);
+    uint256 totalLockCount = tokenLocking.getTotalTokenLockCount(address(token));
+    require(userLockCount < totalLockCount && lastDepositTimestamp < payout.blockTimestamp, "colony-reward-payout-invalid-token-lock");
```

Comment:
You say you've introduced `lastDepositTimestamp` to avoid being able to double spend. Can you explain a scenario where that would be possible?

Reply:
Ok i will try. So we are talking about the implementation without `lastDepositTimestamp`, so in this line above we would just have `require(userLockCount < totalLockCount)`.

So, this will be the timeline of events:
- there is `address1` which have reputation of 10 and 10 tokens, `address2` which also has reputation of 10 and 10 tokens, and `address3` with reputation 10 and 0 tokens
- `address1` and `address2` deposits all of their tokens to token locking contract
- **first** reward payout starts, locking all deposited tokens
- `address1` claims their payout, withdraws their tokens
- `address1` sends all of their tokens to `address3`
- `address3` deposits their tokens (they can because their `lockedBalance` is `0`)
- **second** reward payout starts, locking the tokens of `address3`
- `address3` is now eligable to claim the **first** reward payout since they have 10 reputation, 10 tokens, and their tokens are locked i.e `userLockCount < totalLockCount` is true.
- this can be repeated until funds from reward pot are drained, so `address2` would be the victim here

Reply:
Ok I understand. It's a valid scenario but it feels cleaner if we just not allow any deposits while a token is locked by simply removing that check in `tokenNotLocked` modifier?

Reply:
You mean the check `if (lockedBalances[_token][msg.sender] > 0)` in `tokenNotLocked `?

If we remove it, we would also remove `userTokenLockCount[_token][msg.sender] = totalTokenLockCount[_token]` since it will be useless.

That further means that `lastDepositTimestamps` is also not needed, and we will end up with the very first implementation.

Reply:
Yes that check in `tokenNotLocked` is what I meant. 

The very first implementation did not cater for locking the same token in multiple colonies and also assumed the value of `rewardPayoutId`

Reply:
Yeah, sorry. What i meant is that we will end up with something similar to first implementation. So locking the same token for multiple colonies would remain, but so will the characteristic that we didn't like, which is: tokens will be locked for everybody, regardless if they deposited or not

Reply:
So the main disadvantage is that if someone's balance is `0` they are locked to deposit? Anything else we are sacrificing in this approach?

Reply:
For now, no.

Im just curious how will this approach behave in real world, when bunch of colonies start locking the same token.

I must say that my favourite one was the second one, where the user was responsible for locking, but the counters are also nice.

So, i would like to hear final decision from you and @area also?

Reply:
About this:
>Or when some random colony locks the token that is used in several colonies

The colony only ever locks its _own_ token, even when payouts are triggered for other tokens it works with. For colonies that decide to reuse tokens as their own native currency this may need to be aided in the UI with sufficient gaps in between reward payouts

Reply:
I liked the block timestamp logic in the second implementation too, but I think delegating the responsibility for locking to the user won't work in real life. What we have here, pending the fixes to last review is probably best.